CIP-0144? | Wallet Connector API (#957)
cardano-foundation/CIPs: CIP-0144/README.md

    }
  ]
}
```

Get a specific proposal by id.

##### Era

###### Summary

```
{
  "operation": "getEraBySummary",
  "request": {},
  "response": {
    "type": "object",
    "properties": {
      "summary": {
        "type": "array",
        "items": {
          "$ref": "#/cip-139/EraSummary"
        }
      }
    },
    "required": [
      "summary"
    ]
  },
  "errors": [
    {
      "$ref": "#/appendix/APIError"
    }
  ]
}
```

Get the start and end of each era along with parameters that can vary between hard forks.

### Versioning

In this CIP we are defining two different APIs: the connection API for wallets, and the CIP-30 *[maybe this should have another name to prevent confusion?]* extension which enables an own-data wallet. These two are separate components, below there is a table with separate entries for the versions of the connection API and CIP-30 Extension respectively. 

While the CIP is in preparation, these versions shall be set to `0.0.0`. The moment this CIP is merged the versions shall be set to `1.0.0`, and all implementations should consider that the current version. Any changes to either API should come in form of PRs to this CIP.

> |     |    |
> | --- | --- |
> | Version-Connection-API | 0.0.0 |
> | Version-CIP-30-Extension | 0.0.0 |

Reviewer: ```suggestion
| API | Version |
| --- | --- |
| Version-Connection-API | 0.0.0 |
| Version-CIP-30-Extension | 0.0.0 |
```
... follow up to https://github.com/cardano-foundation/CIPs/pull/957#discussion_r1908650806 & https://github.com/cardano-foundation/CIPs/pull/957/commits/1b4e5b36837d9f67a3cea8052962ea52319eaf69, to either give column headers to this table... or the header bar could be removed by cutting the first two lines.

- A clear separation between the connection mechanism and the different APIs offered by the wallet.
- Using JSON for Cardano domain types instead of CBOR
- Add a query layer API to enable a full data wallet
- Add versioning support to the extension API
- Define this in a transport agnostic way

In its current state, CIP-30 defines it API through a specific transport layer, namely an injected Javascript object.
In this CIP we want to be able to define an API without committing to a specific transport layer. Implementors of this API can choose to support this API through several transports such as: HTTP, an injected Javascript object, JSON-RPC etc.
Furthermore, we want to use JSON-schema to clearly define the types that each method, or operation, expects to receive or returns.
To keep the specification abstract we will use the word "operation" to describe the actions supported by the API: these would map to "endpoints" for an HTTP implementation of the API, or to "methods" for an implementation based on an injected Javascript object.

We will use the following schema to define operations:

> ```

Reviewer 2: ```suggestion
```json
```

{
  "operation": { "type": "string" },
  "request": { "type": "object" },
  "response": { "type": "object" },
  "errors": {
    "type": "array",
    "items": { "type": "object" }
  }
}
```

where `operation` will be the identifier for the operation, `request` and `response` will be the JSON-schemas for the request and response types respectively, and `errors` will be a list of schemas for the errors that the operation may raise.
We do not add an explicit scope to operation names, we do however encourage transport-specific implementations of this API to pick the scoping structure that makes more sense to them.

We will reference several JSON schemas throughout the document, these are:

> - [cip-116](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0116) which provides a JSON encoding of Cardano ledger types. Note that this CIP defines a schema for each ledger era. When referring to a type from this schema we refer to an `anyOf` of all the schemas in which that type is defined. *[Maybe we should only reference the latest era and update the CIP in the future? our current definition requires to be perpetually backward compatible with old ledger types]*

Reviewer 2: ```suggestion
- [CIP-116 | Standard JSON encoding for Domain Types](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0116) which provides a JSON encoding of Cardano ledger types. Note that this CIP defines a schema for each ledger era. When referring to a type from this schema we refer to an `anyOf` of all the schemas in which that type is defined. *[Maybe we should only reference the latest era and update the CIP in the future? our current definition requires to be perpetually backward compatible with old ledger types]*
```
maybe a little nicer for navigation

##### Proposal

###### All

```
{
  "operation": "getAllProposals",
  "request": {},
  "response": {
    "type": "object",
    "properties": {
      "proposals": {
        "type": "array",
        "items": {
          "$ref": "#/cip-139/Proposal"
        }
      }
    },
    "required": [
      "proposals"
    ]
  },
  "errors": [
    {
      "$ref": "#/appendix/APIError"
    }
  ]
}
```

Get all known proposals.

###### Id

```
{
  "operation": "getProposalById",
  "request": {
    "$ref": "#/cip-139/ProposalId"
  },
  "response": {
    "$ref": "#/cip-139/Proposal"
  },
  "errors": [
    {
      "$ref": "#/appendix/APIError"
    }
  ]
}
```

Get a specific proposal by id.

##### Era

###### Summary

```
{
  "operation": "getEraBySummary",
  "request": {},
  "response": {
    "type": "object",
    "properties": {
      "summary": {
        "type": "array",
        "items": {
          "$ref": "#/cip-139/EraSummary"
        }
      }
    },
    "required": [
      "summary"
    ]
  },
  "errors": [
    {
      "$ref": "#/appendix/APIError"
    }
  ]
}
```

Get the start and end of each era along with parameters that can vary between hard forks.

### Versioning

> In this CIP we are defining two different APIs: the connection API for wallets, and the CIP-30 *[maybe this should have another name to prevent confusion?]* extension which enables an own-data wallet. These two are separate components, below there is a table with separate entries for the versions of the connection API and CIP-30 Extension respectively. 

Reviewer 2: I know we already have a lot of CIPs/CPSs for this, and i don't want to increase work on this unnecessarily

It may make sense to have this proposal introduce the connection API, as I believe this API is agnostic of type of wallet(?)
AND then a separate proposal which adds the support/extension for the CIP30 functionality 

does this sound reasonable?

Author: Hey @Ryun1 I think that is a fair point. One argument that can be made against it is that we could split this even further. It is something that I was thinking quite a bit about when writing this document, in a way you could argue that the 5 points listed [here](https://github.com/cardano-foundation/CIPs/pull/957/files#diff-77fa7d71c1149f727dc693fb12eb68c95af075b7cc947bcb6bf9a673682d6ff7R72-R76) should all be standalone CIPs.
Having said this, I think it does make sense to have it separate, if you and other reviewers agree on it I am happy to separate them. My initial thought would be to have a CIP which introduces the transport agnostic specification, and use that to define the connection API. Then we could have one (or two?) other CIPs to define the CIP-30 and CIP-139 extensions. What are your thoughts? cc @rphair

##### Name

```
{
  "operation": "name",
  "request": {},
  "response": { "type": "string" },
  "errors": []
}
```

A name for the wallet which can be used inside of the dApp for the purpose of asking the user which wallet they would like to connect with.

##### Icon

```
{
  "operation": "icon",
  "request": {},
  "response": { "type": "string" },
  "errors": []
}
```

A URI image (e.g. data URI base64 or other) for img src for the wallet which can be used inside of the dApp for the purpose of asking the user which wallet they would like to connect with.

##### ApiVersion

```
{
  "operation": "apiVersion",
  "request": {},
  "response": { "$ref": "#/appendix/SemVer" },
  "errors": []
}
```

Returns the API version for the wallet connection API. This must correspond to the value of `Version-Connection-API` specified in this document, appropriately transformed into a `SemVer` object.

### Versioning

> In this CIP we define a semantic versioning (SemVer) scheme that will be used by the connection API. Extensions defined in other CIPs (e.g. [CIP-XXX | Own data wallet](link)) must follow the same versioning schema.

Author: ```suggestion
In this CIP we define a semantic versioning (SemVer) scheme that will be used by the connection API. Extensions defined in other CIPs (e.g. [CIP-XXX | Own data wallet](https://github.com/cardano-foundation/CIPs/pull/986)) must follow the same versioning schema.
```

- Add an `api.apiVersion` method that returns the version for the connector

Implementing this wrapper is out of scope of this CIP, but we welcome any effort in this direction, as it would make transitioning to this CIP easier for most wallets.

## Rationale: how does this CIP achieve its goals?

The goal of this CIP is to define the standard for a new wallet connector. This connector improves on CIP-30 by defining an independent connection API and giving dApp developers a finer-grained choice on which functionalities to enable via the extension mechanism.
The goal is to be able to clearly differentiate and communicate the type of wallet that is needed, following the distinction made in the [wallet role](https://github.com/cardano-foundation/CIPs/blob/master/CPS-0010/README.md#4-work-within-the-role-of-wallet) section of CPS-10.

There are two CIPs that are connected with this one:

> - [CIP-XXX | Own data wallet](link) which defines an extension for an [own-data wallet](https://github.com/cardano-foundation/CIPs/blob/master/CPS-0010/README.md#own-data-wallets) which is equivalent to what CIP-30 offers today

Author: ```suggestion
- [CIP-XXX | Own data wallet](https://github.com/cardano-foundation/CIPs/pull/986) which defines an extension for an [own-data wallet](https://github.com/cardano-foundation/CIPs/blob/master/CPS-0010/README.md#own-data-wallets) which is equivalent to what CIP-30 offers today
```

---
CIP: 144

> Title: Wallet Connector API

Reviewer 2: ```suggestion
Title: Extensible Wallet Connector Framework
```
I feel maybe we can make the title a bit more specific
This CIP is defining a new framework for wallet connectors, whilst also proposing a connector API

Reviewer: yes, that's good :rocket: but I still think the shorter form (`Wallet Connector -`) should be used for its compliant extensions (e.g. https://github.com/cardano-foundation/CIPs/pull/986#pullrequestreview-2624359783).

Category: Tools
Status: Proposed
Authors:
  - Giovanni Garufi <giovanni@mlabs.city>
Implementors: []
Discussions:
    - https://github.com/cardano-foundation/cips/pulls/957
Created: 2024-12-13
License: CC-BY-4.0

> Solution-To: CPS-0010

Reviewer: yes please... but without the hyphen 😅 as in this pending example:
* https://github.com/cardano-foundation/CIPs/pull/986

---

## Abstract

CIP-30 is the standard interface of communication between wallets and dApps. While this CIP has been instrumental in the development of dApps for Cardano, it also has some shortcomings that have been observed across several implementations.

[CPS-0010](https://github.com/cardano-foundation/CIPs/blob/master/CPS-0010/README.md) outlines the shortcomings of CIP-30. In this CIP we aim to address some of the issues pointed out by CPS-10: splitting up the connection API from the functionality provided by the [full API](https://cips.cardano.org/cip/CIP-30#full-api), defining the API in a transport agnostic way and defining a versioning mechanism for the connection API and its extensions.

## Motivation: why is this CIP necessary?

CIP-30 is a universally accepted web-based wallet standard for Cardano. It provides a minimalistic interface that, in principle, can be used to build almost any kind of Cardano dApp. However, the way dApp<->wallet interaction is defined leads to suboptimal dApp architecture due to CIP-30 limits.

### CPS-10

> [CPS-0010](https://github.com/Ryun1/CIPs/blob/cps-wallet-connector/CPS-0010/README.md) discusses several limitations of CIP-30. In this CIP we try to solve the issues pointed out in that CPS. There are still some areas to improve on: things like the event listener API are intentionally left out of this CIP to prevent bloat of scope. We welcome future CIPs, or updates to this CIP to refine any limitation that is not tackled in this document.

Reviewer 3: I was thinking that extensions:
- would be defining their schemas for base operation payloads (`request`/`response`/`error`)
- probably shouldn't be adding new base operations, as it adds new requirements for transport CIPs/implementations

Given this CIP (and naming of `request`/`response` operations), transport specifications would probably imply only a single value resolved from a 'request' - a typical pull-based API structure, e.g. injected javascript object transport implementation would probably map requests/responses into a promise based API like `api.cip123.operation(r: Request) => Promise<Response>`

If you wanted to add an extension for events, you couldn't really do it by adding an `on` 'request', because transports would generically expect it to resolve with only a single value like all other requests.

Am I missing something - could you create an extension for events given this base CIP?

- A clear separation between the connection mechanism and the different APIs offered by the wallet.
- Using JSON for Cardano domain types instead of CBOR
- Add versioning support to the extension API
- Define the API in a transport agnostic way

### Transport agnostic operations

In its current state, CIP-30 defines it API through a specific transport layer, namely an injected Javascript object.
In this CIP we want to be able to define an API without committing to a specific transport layer. Implementors of this API can choose to support this API through several transports such as: HTTP, an injected Javascript object, JSON-RPC etc.
Furthermore, we want to use JSON-schema to clearly define the types that each method, or operation, expects to receive or returns.
To keep the specification abstract we will use the word "operation" to describe the actions supported by the API: these would map to "endpoints" for an HTTP implementation of the API, or to "methods" for an implementation based on an injected Javascript object.

We will use the following schema to define operations:

```json

> {

Reviewer 2: ```suggestion
{
  "connectionVersion": { "type": "semVar" },
```

does it make sense to have the connection API versioned?

Author: I think so, I added an endpoint both to the connection API and to the different extensions to report their `apiVersion`. Adding it here however, would potentially make each endpoint have it's own separate version which might be a bit too much complexity

- [appendix](#appendix) in which we define schemas for types required by the connection API and the error types

We will use an identifier in the anchor to refer to the schema where each type is defined. For example, if we want to reference the `Transaction` type, as defined in CIP-116 we will use the following schema reference `{ "$ref": "#/cip-116/Transaction" }`.

If an operation does not require any arguments as part of it's request, or does not return a meaningful response, we will represent this as an `{}` on the respective operation field. The arguments to an operation will always be represented as an object. If a field is not marked as `required`, then that argument is to be considered optional.

For each operation we will provide some details on how the implementation should behave. Note that some of these are taken verbatim from CIP-30.

### Connection API

The role of the connection API is to provide generic information about the wallet and to allow the user to opt into the functionalities (or extensions) that they want the wallet to provide.

##### Enable

> ```

Reviewer 2: ```suggestion
```json
```

Author: I'll add this across all CIPs

Reviewer 2: awesome thanks!

{
  "operation": "enable",

>   "request": { "$ref": "#/appendix/Extensions" },

Reviewer 2: just a thought, would there be a usecase where a dApp doesnt request a whole extension?
perhaps a dApp only wants a couple methods, maybe this would just add unneeded complexity

Reviewer 2: Discussed in meeting
we can consider this in the future, or when we have more feedback from stakeholders

>   "request": { "$ref": "#/appendix/Extensions" },

Reviewer 2: i feel that one Cardano network at a time, is probably okay
and matches existing implementations

>   "request": { "$ref": "#/appendix/Extensions" },

Reviewer 3: Agree, one network should be enough for now. Although it would be interesting to learn about Hydra use cases

  "response": {},
  "errors": [
    { "$ref": "#/appendix/APIError" }
  ]
}
```

> This is the entrypoint to start communication with the user's wallet. The wallet should request the user's permission to connect the web page to the user's wallet, and if permission has been granted, the APIs will be made available for the dApp to use. The wallet can choose to maintain a whitelist to not necessarily ask the user's permission every time access is requested, but this behavior is up to the wallet and should be transparent to web pages using this API. If a wallet is already connected this function should not request access a second time.

Reviewer 2: > but this behavior is up to the wallet and should be transparent to web pages using this API.

I like this a lot, this gives wallets the freedom to decide

Through the `extension` field, dApps can request a list of what functionality they expect as a list of CIP numbers (and optional versions) capturing those extensions. This is used as an extensibility mechanism to document what functionalities can be provided by the wallet interface. We will see later in this document examples of what such extensions might be and which functionalities they will enable. New functionalities can be introduced via additional CIPs and may be all or partially supported by wallets.

When requesting the functionalities of a cip extension, dApps can optionally specify a version number for it. Every cip extension defined in the future, must also define a versioning scheme following SemVer. When the version argument is not specified, wallets should take that as the greater version they implement. In this context - and for the rest of this discussion - we will assume versions are ordered with the canonical ordering of SemVer. If the dApp requests a specific version of an extension, wallets can only accept the request if a version of the extension they implement has:

> - the same major value as the one requested by the dApp AND

Reviewer 2: do we need to include SemVer versioning for extensions?
I feel like this maybe adding unneeded complexity

Existing CIP30 extensions, aren't versioned, instead to replace one you must provide another CIP
I think this is mostly fine

Reviewer: In prior discussions for both CIPs and CIP-0030 extensions for last 4 years or so, in my recollection all parties agreed each time that an integer "serial number" was both adequate and future-proof as a version ID _except_ for those who argued that "semantic versioning is _always_ preferred" as a general principle.

Reviewer 3: I think we should have the option to update an API through versions. It enables a more 'agile' approach to defining and implementing CIPs

DApps are expected to use this endpoint to perform an initial handshake and ensure that the wallet supports all their required functionalities. Note that it's possible for two extensions to be mutually incompatible (because they provide two conflicting features). While we may try to avoid this as much as possible while designing CIPs, it is also the responsibility of wallet providers to assess whether they can support a given combination of extensions, or not. Wallets should throw an error with code `NotSatisfiable` if either they do not support the required extension, or they deem the combination of extensions requested by the dApp to be incompatible. In this case, it is up to the dApp to decide to retry with different requirements, or to give up on establishing a connection with the wallet.

##### IsEnabled

```
{
  "operation": "isEnabled",
  "request": {},
  "response": { "type": "boolean" },
  "errors": [
    { "$ref": "#/appendix/APIError" }
  ]
}
```

Returns true if the dApp is already connected to the user's wallet, or if requesting access would return true without user confirmation (e.g. the dApp is whitelisted), and false otherwise.

> ##### SupportedExtensions

Reviewer 2: does connection have to be enabled to be able to call this method?

##### Name

```
{
  "operation": "name",
  "request": {},
  "response": { "type": "string" },
  "errors": []
}
```

A name for the wallet which can be used inside of the dApp for the purpose of asking the user which wallet they would like to connect with.

##### Icon

```
{
  "operation": "icon",
  "request": {},
  "response": { "type": "string" },
  "errors": []
}
```

A URI image (e.g. data URI base64 or other) for img src for the wallet which can be used inside of the dApp for the purpose of asking the user which wallet they would like to connect with.

##### ApiVersion

```
{
  "operation": "apiVersion",
  "request": {},
  "response": { "$ref": "#/appendix/SemVer" },
  "errors": []
}
```

Returns the API version for the wallet connection API. This must correspond to the value of `Connection-API` specified in this document, appropriately transformed into a `SemVer` object.

### Versioning

> 

Reviewer 2: note to discuss using SemVar or other versioning scheme

In this CIP we define a semantic versioning (SemVer) scheme that will be used by the connection API. Extensions defined in other CIPs (e.g. [CIP-XXX | Own data wallet](https://github.com/cardano-foundation/CIPs/pull/986)) must follow the same versioning schema.

While the CIP is in preparation, the version shall be set to `0.0.0`. The moment this CIP is merged the version shall be set to `1.0.0`, and all implementations should consider that the current version. Any changes to the connection API should come in form of PRs to this CIP.

| API | Version |
| --- | --- |
| Connection-API | 0.0.0 |

### Appendix

This appendix contains additional schemas for types that are used in the APIs.

> #### Connection API Data Types

Reviewer 2: really nice idea to include these :)

    "info": {
      "type": "string"
    }
  },
  "required": ["code"],
}
```

- `InvalidRequest`: (-1) Inputs do not conform to this spec or are otherwise invalid.
- `InternalError`: (-2) An error occurred during execution of this API call.
- `Refused`: (-3) The request was refused due to lack of access - e.g. wallet disconnects.
- `AccountChange`: (-4) The account has changed. The dApp should call wallet.enable() to reestablish connection to the new account. The wallet should not ask for confirmation as the user was the one who initiated the account change in the first place.
- `NotSatisfiable`: (-5) The request is structurally correct, but the wallet can not satisfy it for some reason.

#### Transport specific connectors

While this CIP attempts to define an API in a transport agnostic way, implementations will be forced to pick a specific transport. In the spec we have not given specifics on how to namespace and generally structure the api in an implementation. This is both because details depend on the underlying transport that is chosen to implement the API, but also because we wanted to make backwards compatibility with the original CIP-30 proposal as easy as possible.

> Each transport specific implementation will make a choice on how to namespace access to the operations. In the following we give a description of how this should work for an implementation using an injected javascript object. Support for different transports can be added to this CIP in form of PRs.

Reviewer: @mkazlauskas @nazrhom this is a question I would hope to have @Ryun1's insight about.

> Each transport specific implementation will make a choice on how to namespace access to the operations. In the following we give a description of how this should work for an implementation using an injected javascript object. Support for different transports can be added to this CIP in form of PRs.

Reviewer 2: > The argument against this would be to avoid opening too many interconnected CIPs

i feel that is probably okay
since it would be quite nice if this CIP was transport-agnostic

##### APIError

```
{
  "type": "object",
  "title": "APIError",
  "properties": {
    "code": {
      "type": "number",
      "title": "APIErrorCode"
      "enum": [-1, -2, -3, -4, -5]
    },
    "info": {
      "type": "string"
    }
  },
  "required": ["code"],
}
```

- `InvalidRequest`: (-1) Inputs do not conform to this spec or are otherwise invalid.
- `InternalError`: (-2) An error occurred during execution of this API call.
- `Refused`: (-3) The request was refused due to lack of access - e.g. wallet disconnects.
- `AccountChange`: (-4) The account has changed. The dApp should call wallet.enable() to reestablish connection to the new account. The wallet should not ask for confirmation as the user was the one who initiated the account change in the first place.
- `NotSatisfiable`: (-5) The request is structurally correct, but the wallet can not satisfy it for some reason.

> 

Reviewer 2: would a new error be needed 
for the case where a dApp requests a network at enable time
but the wallet is not connected to it?
